File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function solution(n, wires) {
    let minDiff = Infinity;

    const buildGraph = (skipIdx) => {
        const graph = Array.from({ length: n + 1 }, () => []);
        wires.forEach(([a, b], idx) => {
            if (idx === skipIdx) return; // 간선 하나 끊기
            graph[a].push(b);
            graph[b].push(a);
        });
        return graph;
    };

    const countNodes = (start, graph, visited) => {
        let count = 0;
        const queue = [start];
        visited[start] = true;

        while (queue.length) {
            const node = queue.shift();
            count++;

            graph[node].forEach((neighbor) => {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    queue.push(neighbor);
                }
            });
        }

        return count;
    };

    for (let i = 0; i < wires.length; i++) {
        const graph = buildGraph(i);
        const visited = Array(n + 1).fill(false);
        const count = countNodes(1, graph, visited); // 한 쪽 트리의 노드 수
        const diff = Math.abs(n - 2 * count); // 두 쪽의 노드 수 차이
        minDiff = Math.min(minDiff, diff);
    }

    return minDiff;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    graph = [[] for _ in range(n + 1)]
    
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)
    
    def dfs(node, visited, cut_a, cut_b):
        visited[node] = True
        count = 1
        
        for next_node in graph[node]:
            # 끊은 전선이면 무시
            if (node == cut_a and next_node == cut_b) or (node == cut_b and next_node == cut_a):
                continue
                
            if not visited[next_node]:
                count += dfs(next_node, visited, cut_a, cut_b)
        
        return count
    
    answer = n
    
    for a, b in wires:
        visited = [False] * (n + 1)
        
        # 아무 노드 하나에서 시작해서 연결된 개수 세기
        count = dfs(1, visited, a, b)
        diff = abs(count - (n - count))
        answer = min(answer, diff)
    
    return answer